File: 152. Maximum Product Subarray/java/Solution.java
class Solution {
    public int maxProduct(int[] nums) {
        int min = 1, max = 1, ans = Integer.MIN_VALUE;

        for (int num: nums) {
            int temp1 = num * min, temp2 = num * max;
            min = min(num, temp1, temp2);
            max = max(num, temp1, temp2);

            ans = Math.max(ans, max);
        }

        return ans;
    }

    int min(int a, int b, int c) {
        return Math.min(a, Math.min(b, c));
    }

    int max(int a, int b, int c) {
        return Math.max(a, Math.max(b, c));
    }
}

File: 152. Maximum Product Subarray/rust/solution.rs
struct Solution;

impl Solution {
    pub fn max_product(nums: Vec<i32>) -> i32 {
        let (mut min, mut max, mut ans) = (1, 1, 0);

        for num in nums {
            let (a, b) = (min * num, max * num);
            min = i32::min(num, i32::min(a, b));
            max = i32::max(num, i32::max(a, b));

            ans = i32::max(ans, max);
        }

        ans
    }
}